chain/chunks/src/lib.rs

```diff
@@ -1075,6 +1092,139 @@ impl ShardsManager {
         )
     }
 
+    /// Constructs a vector of [`ReceiptsProof`]s for specified chunk.
+    fn make_outgoing_receipts_proofs<T>(
+        &self,
+        chunk_header: &ShardChunkHeader,
+        outgoing_receipts: &Vec<Receipt>,
+        make_elem: impl Fn(ReceiptProof) -> T,
+    ) -> Result<Vec<T>, near_chunks_primitives::Error> {
```

Comment:
My personal style is to write such things as

```suggestion
        sink: &mut dyn FnMut(ReceiptProof),
    ) -> Result<(), near_chunks_primitives::Error> {
```

That is, without generics. The primary motivation is compile time -- with generics, this closure will be "copy pasted" several times, and it feels like it can add up other time.

I can not definitely say that using vs not-using such pattern would lead to some definitive X difference in compile time.

However, in rust-analyzer we did get observable and measurable compile-time improments out of this style of changes.

For example, https://github.com/rust-analyzer/rust-analyzer/commit/76b60efbfb8cdc94cf14fef1be50c60f250bc293 shows about 10% percent improvement in terms of amount of LLVM IR generated, when we replaced one generic call-back with a dynamic call-back (in the context of the codebase which uses generics rather sparingly)

Reply:
It's definitely not a clear win, it's a trade off between "write more code" and "wait more for the code to compile", and its indeed unclear how much of the second one we gain/loose here, so no strong opinion. 

But, having taken a second look here, it seems that this can just return an iterator? That seems like a clear win -- no geneircs, no closures, more flexibility to the caller:

```diff
diff --git a/chain/chunks/src/lib.rs b/chain/chunks/src/lib.rs
index cb55bac6a..247684ad9 100644
--- a/chain/chunks/src/lib.rs
+++ b/chain/chunks/src/lib.rs
@@ -1097,12 +1097,11 @@ impl ShardsManager {
     /// The `make_elem` argument converts `ReceiptsProof` values into objects
     /// which are returned from this function.  In practice this is used to
     /// either return `ReceiptProof` directly or wrap the object into `Arc`.
-    fn make_outgoing_receipts_proofs<T>(
+    fn make_outgoing_receipts_proofs(
         &self,
         chunk_header: &ShardChunkHeader,
         outgoing_receipts: &[Receipt],
-        make_elem: impl Fn(ReceiptProof) -> T,
-    ) -> Result<Vec<T>, near_chunks_primitives::Error> {
+    ) -> Result<impl Iterator<Item = ReceiptProof>, near_chunks_primitives::Error> {
         let shard_id = chunk_header.shard_id();
         let shard_layout = self
             .runtime_adapter
@@ -1119,20 +1118,13 @@ impl ShardsManager {
 
         let mut receipts_by_shard =
             Chain::group_receipts_by_shard(outgoing_receipts.to_vec(), &shard_layout);
-        let proofs = proofs
-            .into_iter()
-            .enumerate()
-            .map(|(proof_shard_id, proof)| {
-                let proof_shard_id = proof_shard_id as u64;
-                let receipts = receipts_by_shard.remove(&proof_shard_id).unwrap_or_else(Vec::new);
-                let shard_proof = ShardProof {
-                    from_shard_id: shard_id,
-                    to_shard_id: proof_shard_id,
-                    proof: proof,
-                };
-                make_elem(ReceiptProof(receipts, shard_proof))
-            })
-            .collect();
+        let proofs = proofs.into_iter().enumerate().map(move |(proof_shard_id, proof)| {
+            let proof_shard_id = proof_shard_id as u64;
+            let receipts = receipts_by_shard.remove(&proof_shard_id).unwrap_or_else(Vec::new);
+            let shard_proof =
+                ShardProof { from_shard_id: shard_id, to_shard_id: proof_shard_id, proof: proof };
+            ReceiptProof(receipts, shard_proof)
+        });
         Ok(proofs)
     }
 
@@ -1165,11 +1157,12 @@ impl ShardsManager {
         // proofs.
         let outgoing_receipts = chunk.receipts();
         let receipt_proofs = self
-            .make_outgoing_receipts_proofs(&header, &outgoing_receipts, |p| p)
+            .make_outgoing_receipts_proofs(&header, &outgoing_receipts)
             .map_err(|err| {
                 warn!(target: "chunks", "Not sending {}, {}", request.chunk_hash.0, err);
             })
-            .ok()?;
+            .ok()?
+            .collect::<Vec<_>>();
 
         // Construct EncodedShardChunk.  If we earlier determined that we will
         // need parity parts, instruct the constructor to calculate them as
@@ -2064,7 +2057,8 @@ impl ShardsManager {
         store_update: &mut ChainStoreUpdate<'_>,
     ) -> Result<(), Error> {
         let header = encoded_chunk.cloned_header();
-        let receipts = self.make_outgoing_receipts_proofs(&header, &outgoing_receipts, |p| p)?;
+        let receipts =
+            self.make_outgoing_receipts_proofs(&header, &outgoing_receipts)?.collect::<Vec<_>>();
         let partial_chunk = PartialEncodedChunkV2 {
             header,
             parts: encoded_chunk
@@ -2115,8 +2109,10 @@ impl ShardsManager {
             entry.push(part_ord);
         }
 
-        let receipt_proofs =
-            self.make_outgoing_receipts_proofs(&chunk_header, &outgoing_receipts, Arc::new)?;
+        let receipt_proofs = self
+            .make_outgoing_receipts_proofs(&chunk_header, &outgoing_receipts)?
+            .map(Arc::new)
+            .collect::<Vec<_>>();
         for (to_whom, part_ords) in block_producer_mapping {
             let part_receipt_proofs = receipt_proofs
                 .iter()
```

Reply:
Well, yeah, I considered that and that works for me but notice that…

> That seems like a clear win -- no geneircs,

…there are generics here.

Reply:
There's no generics here, `impl Trait` syntax does very different things before and after `->`

```rust
// This is generic: for every distinct type of `x`, we'll get a separate copy of `generic`: `generic<i32>`, `generic<Vec<()>>`, etc
fn generic(x: &impl Clone) {
  let _x = x.clone();
}


// This is not generic -- there's only one copy of `monomorphic`, it return a single type (`u32`), it's just that we are too lazy to spell it out
fn monomorphic() -> impl Clone {
  92u32
}
```

Reply:
Ah, yeah, you’re right.